### ISSUE TITLE
Axum 7.x causes compilation errors. 

### ISSUE THREAD

**Issue by dipthegeezer** [awslabs/aws-lambda-rust-runtime#738]
Hi 

I'm fairly new to Rust, so please advise if I'm doing something wrong. I was trying to get the [http-axum](https://github.com/awslabs/aws-lambda-rust-runtime/tree/main/examples/http-axum) example working and wanted to use the latest package dependencies. However axum 7.x no longer works - they have just [released](https://github.com/tokio-rs/axum/pull/1751) a braking change which I guess is the issue.  

I got the following error:

```
--> crates/xxxxxx/src/main.rs:45:9
    |
45  |     run(app).await
    |     --- ^^^ the trait `Service<lambda_http::http::Request<lambda_http::Body>>` is not implemented for `Router<_>`
    |     |
    |     required by a bound introduced by this call
    |
    = help: the following other types implement trait `Service<Request>`:
              <Router as Service<axum::http::Request<B>>>
              <Router as Service<IncomingStream<'_>>>
note: required by a bound in `lambda_http::run`
   --> /Users/xxxx/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lambda_http-0.8.3/src/lib.rs:191:8
    |
189 | pub async fn run<'a, R, S, E>(handler: S) -> Result<(), Error>
    |              --- required by a bound in this function
190 | where
191 |     S: Service<Request, Response = R, Error = E>,
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `run`
```

My Cargo file dependency looks like this:

```
[dependencies]
lambda_http = "0.8.3"
lambda_runtime = "0.8.3"
tokio = { version = "1.34.0", features = ["macros"] }
tracing = { version = "0.1.40", features = ["log"] }
tracing-subscriber = { version = "0.3.18", default-features = false, features = ["fmt"] }
axum = "0.7.1"
serde_json = "1.0.108"
```

I have downgraded in the meantime to stick to the example given but I thought I would bring this to your attention.

**Comment by simbleau** [comment]
Also having this issue.
<img width="1408" alt="image" src="https://github.com/awslabs/aws-lambda-rust-runtime/assets/48108917/8ca9d531-00bd-4472-900a-6f7d60ebf68e">

**Comment by Torsitano** [comment]
> This PR should fix it, but I'll really appreciate you testing your function with it before I merge it:
> 
> #749
> 
> You can install the dependency from GitHub directly using this in your Cargo.toml file:
> 
> ```
> lambda_http = { git = "https://github.com/awslabs/aws-lambda-rust-runtime", branch = "hyper1_upgrade" }
> ```

Hey @calavera, I really appreciate the work you put into this. Unfortunately I'm still running into the same issue with `run(app)`.

If it helps narrow it down, the `run` function is in `lambda-http/src/lib.rs`. I'm getting the same error while using the branch version as I did with the current release:
```
error[E0277]: the trait bound `Router<Backend>: Service<axum::http::Request<lambda_http::Body>>` is not satisfied
   --> src/main.rs:75:9
    |
75  |     run(app).await
    |     --- ^^^ the trait `Service<axum::http::Request<lambda_http::Body>>` is not implemented for `Router<Backend>`
    |     |
    |     required by a bound introduced by this call
    |
    = help: the following other types implement trait `Service<Request>`:
              <Router as Service<axum::http::Request<B>>>
              <Router as Service<IncomingStream<'_>>>
note: required by a bound in `lambda_http::run`
   --> /home/shawn/.cargo/git/checkouts/aws-lambda-rust-runtime-7c865cce90132439/6f1a457/lambda-http/src/lib.rs:191:8
    |
189 | pub async fn run<'a, R, S, E>(handler: S) -> Result<(), Error>
    |              --- required by a bound in this function
190 | where
191 |     S: Service<Request, Response = R, Error = E>,
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `run`

```

**Comment by Torsitano** [comment]
@calavera I'd be happy to, let me setup a basic example demonstrating it(just to prevent additional complications from my convoluted poorly written code), and I'll send it over ASAP.

**Comment by Torsitano** [comment]
Got it, there's a strong chance the issue is me. Which I'm also super happy to find, since that's still a solution.

**Comment by dipthegeezer** [comment]
> This PR should fix it, but I'll really appreciate you testing your function with it before I merge it:
> 
> #749
> 
> You can install the dependency from GitHub directly using this in your Cargo.toml file:
> 
> ```
> lambda_http = { git = "https://github.com/awslabs/aws-lambda-rust-runtime", branch = "hyper1_upgrade" }
> ```

Thanks @calavera this seems to work fine for me. Though I was not doing anything too far off from the example.

**Comment by Torsitano** [comment]
**TL;DR:** I have no idea what I'm doing and this is above my head to understand what's going on.
-------------

@calavera Looks like there some weird stuff going on, likely from me piecing a bunch of different things together as I'm trying to figure things out.

It seemed like depending on how I was building my routers, sometimes the inferred type would be `Router<{unknown}>`, sometimes `Router<Backend>`(from my state stuff), and sometimes just `Router`.

There were situations where the inferred type would be `Router<_>`(different generics depending on the case) up until the specific object was used in `run()`, ie:
```rust
    // inferred as Router<{unknown}>
    let router = Router::new().route("/health_check", get(routes::health_check));

    // inferred as Router
    let second_router = Router::new().route("/other_check", get(routes::other_check));

    // Swapping which one is used here will swap the inferred types above
    run(second_router).await
```

I ultimately solved the issue in my original code by attaching state/tracing at a different point, so I think the error was occurring from the way that I was building/merging the Routers. I've tried to recreate that with a more basic example, but it seems to be working there, so there's some combination of the middleware/layers/who knows what that was causing some weird inference in my real code that was causing the issue.
